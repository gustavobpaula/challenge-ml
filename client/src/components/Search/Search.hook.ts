import { useCallback } from 'react'
import { useHistory } from 'react-router-dom'
import { useStore } from 'hooks/useStore'

export const useSearch = () => {
  const { query, setQuery } = useStore(state => state)

  const history = useHistory()

  const handleSubmit = useCallback(
    (e: React.FormEvent<HTMLFormElement>) => {
      e.preventDefault()

      if (query) {
        history.push({
          pathname: '/items',
          search: `?q=${query}`,
        })
      }
    },
    [query]
  )

  const handleChange = useCallback(
    (e: React.FormEvent<HTMLInputElement>) => {
      setQuery(e.currentTarget.value)
    },
    [setQuery]
  )

  return {
    query,
    handleChange,
    handleSubmit,
  }
}
