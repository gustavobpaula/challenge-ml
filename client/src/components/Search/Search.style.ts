import styled from '@emotion/styled'

export const Form = styled.form`
  display: flex;
  margin-left: 25px;
  width: 100%;
`

export const Input = styled.input`
  border: none;
  border-radius: 4px 0 0 4px;
  color: ${({ theme }) => theme.colors.gray1};
  font-size: ${({ theme }) => theme.type.sizes.md};
  outline: none;
  padding: 7px 15px 9px 15px;
  width: 100%;
`

export const Button = styled.button`
  align-items: center;
  background-color: ${({ theme }) => theme.colors.gray4};
  border: none;
  cursor: pointer;
  display: flex;
  height: 37px;
  justify-content: center;
  width: 37px;
`
