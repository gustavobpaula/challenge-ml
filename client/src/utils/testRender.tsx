import React from 'react'
import { render, RenderOptions } from '@testing-library/react'
import { ThemeProvider } from '@emotion/react'

import theme from 'config/theme'

type RenderComponentProps = {
  component: React.ReactNode
  useTheme?: boolean
  options?: RenderOptions
}

type ProviderProps = {
  children: React.ReactNode
}

const Theme = ({ children }: ProviderProps) => (
  <ThemeProvider theme={theme}>{children}</ThemeProvider>
)

export const RenderComponent = ({
  component,
  useTheme = false,
  options = {},
}: RenderComponentProps) => {
  const providers = [useTheme && Theme]

  const componentWithProviders = providers.reduce((accComponent, Provider) => {
    if (!Provider) return accComponent

    return <Provider>{accComponent}</Provider>
  }, component)

  return render(<>{componentWithProviders}</>, options)
}
