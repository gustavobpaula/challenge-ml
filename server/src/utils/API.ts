import fetch from 'node-fetch'
import { URLSearchParams } from 'url'

const ML_API = 'https://api.mercadolibre.com'

const API = {
  /**
   * Return currency object
   */
  getCurrency: async (id: string) => {
    try {
      const response = await fetch(`${ML_API}/currencies/${id}`)
      return await response.json()
    } catch (error) {
      throw new Error(error)
    }
  },
  /**
   * Return items by query
   */
  search: async (query: string) => {
    const params = new URLSearchParams()
    params.append('q', query)

    try {
      const response = await fetch(
        `${ML_API}/sites/MLA/search?${params.toString()}`
      )

      return await response.json()
    } catch (error) {
      throw new Error(error)
    }
  },
}

export default API
