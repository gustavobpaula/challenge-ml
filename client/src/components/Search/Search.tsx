import iconSearch from 'assets/ic_Search.png'
import * as S from './Search.style'

function Search() {
  return (
    <S.Form>
      <S.Input type="text" placeholder="Nunca dejes de buscar" />
      <S.Button>
        <img src={iconSearch} alt="Buscar" />
      </S.Button>
    </S.Form>
  )
}

export default Search
