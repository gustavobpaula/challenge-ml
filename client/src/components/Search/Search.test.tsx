import { screen } from '@testing-library/react'
import userEvent from '@testing-library/user-event'
import { renderHook } from '@testing-library/react-hooks'

import { RenderComponent } from 'utils'
import Search from './Search'
import { useSearch } from './Search.hook'

const makeSut = () => {
  return RenderComponent({
    component: <Search />,
    useTheme: true,
    useRouter: true,
  })
}

describe('Search component', () => {
  it('Should render a input', () => {
    makeSut()

    expect(
      screen.getByRole('textbox', { name: /campo de búsqueda/i })
    ).toBeInTheDocument()
  })

  it('Input should change value', () => {
    makeSut()

    const input = screen.getByRole('textbox', {
      name: /campo de búsqueda/i,
    }) as HTMLInputElement

    userEvent.type(input, 'Iphone 64gb')

    expect(input.value).toBe('Iphone 64gb')
  })

  it('Should render a button', () => {
    makeSut()

    const button = screen.getByRole('button', { name: /buscar/i })

    expect(button).toBeInTheDocument()
    expect(button).toBeEnabled()
  })

  it('Should change inputValue on click on Button', () => {
    makeSut()

    const { result } = renderHook(() => useSearch())

    userEvent.type(
      screen.getByRole('textbox', { name: /campo de búsqueda/i }),
      'Iphone 32GB'
    )

    const button = screen.getByRole('button', { name: /buscar/i })

    userEvent.click(button)

    expect(result.current.inputValue).toBe('Iphone 32GB')
  })
})
