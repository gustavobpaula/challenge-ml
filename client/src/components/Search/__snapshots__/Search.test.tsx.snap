// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search component Should render match snapshot 1`] = `
<DocumentFragment>
  <form
    class="css-o9qfam"
  >
    <input
      aria-label="campo de bÃºsqueda"
      class="css-3300kj"
      name="query"
      placeholder="Nunca dejes de buscar"
      type="text"
      value=""
    />
    <button
      class="css-1k0obmh"
      type="submit"
    >
      <img
        alt="Buscar"
        src="ic_Search.png"
      />
    </button>
  </form>
</DocumentFragment>
`;
